File: app/src/main/java/org/ametro/ui/navigation/SchemeNavigationListComparator.java
package org.ametro.ui.navigation;

import org.ametro.model.entities.MapMetadata;

import java.util.Comparator;

public class SchemeNavigationListComparator implements Comparator<MapMetadata.Scheme> {
    @Override
    public int compare(MapMetadata.Scheme lhs, MapMetadata.Scheme rhs) {
        int byType = getType(lhs).compareTo(getType(rhs));
        if(byType != 0){
            return byType;
        }
        return getDisplayName(lhs).compareTo(getDisplayName(rhs));
    }

    private String getType(MapMetadata.Scheme scheme){
        String type = scheme.getTypeName().equals("ROOT") ? scheme.getDisplayName() : scheme.getTypeDisplayName();
        if(scheme.getName().equals("metro") || scheme.getTypeName().equals("Метро")){
            return "";
        }
        if(type.equals("OTHER")){
            return ""+Character.MAX_VALUE;
        }
        return type;
    }

    private String getDisplayName(MapMetadata.Scheme scheme){
        return scheme.getTypeName().equals("ROOT") ? "" : scheme.getDisplayName();
    }

}

File: app/src/main/java/io/github/romangolovanov/apps/ametro/ui/navigation/SchemeNavigationListComparator.kt
package io.github.romangolovanov.apps.ametro.ui.navigation

import io.github.romangolovanov.apps.ametro.model.entities.MapMetadata

class SchemeNavigationListComparator : Comparator<MapMetadata.Scheme> {

    override fun compare(lhs: MapMetadata.Scheme, rhs: MapMetadata.Scheme): Int {
        val byType = getType(lhs).compareTo(getType(rhs))
        if (byType != 0) {
            return byType
        }
        return getDisplayName(lhs).compareTo(getDisplayName(rhs))
    }

    private fun getType(scheme: MapMetadata.Scheme): String {
        val type = if (scheme.typeName == "ROOT") scheme.displayName ?: "" else scheme.typeDisplayName ?: ""
        if (scheme.name == "metro" || scheme.typeName == "Метро") {
            return ""
        }
        if (type == "OTHER") {
            return "" + Char.MAX_VALUE
        }
        return type
    }

    private fun getDisplayName(scheme: MapMetadata.Scheme): String {
        return if (scheme.typeName == "ROOT") "" else scheme.displayName ?: ""
    }
}
